### PR TITLE
Add DNA Center Integration

### PR DESCRIPTION
**Pull request by jdrew82**
This PR is to add the relevant parts for a DNA Center integration.

### CODE REVIEW

File: nautobot_ssot/integrations/dna_center/constants.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+"""Defines various constants used throughout the App."""
+
+from django.conf import settings
+
+# Import config vars from nautobot_config.py
+PLUGIN_CFG = settings.PLUGINS_CONFIG["nautobot_ssot"]
+
+DNAC_PLATFORM_MAPPER = {
```
Comment (Kircheneer) [inline nautobot_ssot/integrations/dna_center/constants.py:8]: Can we use those from here? https://github.com/networktocode/netutils/pull/519

File: nautobot_ssot/integrations/dna_center/diffsync/adapters/nautobot.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+        self.job = job
+        self.sync = sync
+        self.tenant = tenant
+        self.objects_to_create = defaultdict(list)
+        self.objects_to_delete = defaultdict(list)
+
+    def load_regions(self):
+        """Load Region data from Nautobt into DiffSync models."""
+        try:
+            locations = OrmLocation.objects.filter(location_type=self.locationtype_map["Region"])
```
Comment (Kircheneer) [inline nautobot_ssot/integrations/dna_center/diffsync/adapters/nautobot.py:90]: I think this adapter could benefit a lot from using `select_related` and `prefetch_related`. Currently, a lot of SQL queries may be generated when this runs on a big database.

File: nautobot_ssot/integrations/dna_center/diffsync/models/base.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    """DiffSync model for DNA Center areas."""
+
+    _modelname = "area"
+    _identifiers = ("name", "parent")
+    _attributes = ()
+    _children = {}
+
+    name: str
+    parent: Optional[str] = None
+    buildings: Optional[List["Building"]] = []
```
Comment (Kircheneer) [inline nautobot_ssot/integrations/dna_center/diffsync/models/base.py:18]: Is it on purpose that hit is neither identifier, nor attribute or child?

File: nautobot_ssot/integrations/dna_center/utils/dna_center.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        Returns:
+            tuple: Tuple of Site address and type.
+        """
+        address = ""
+        site_type = ""
+        for element in info:
+            if element["nameSpace"] == "Location":
+                address = element["attributes"]["address"]
+                site_type = element["attributes"]["type"]
+        return (address, site_type)
```
Comment (Kircheneer) [inline nautobot_ssot/integrations/dna_center/utils/dna_center.py:77]:
```suggestion
        for element in info:
            if element["nameSpace"] == "Location":
                return (element["attributes"]["address"], element["attributes"]["type"]) 
        return ("", "")
```

A bit more performant

Comment (Kircheneer) [reply nautobot_ssot/integrations/dna_center/utils/dna_center.py:77]: This can also be applied further down

File: pyproject.toml [hunk 5]
```diff
@@ -50,6 +50,7 @@ requests = { version = ">=2.21.0", optional = true }
 requests-oauthlib = { version = ">=1.3.0", optional = true }
 six = { version = ">=1.13.0", optional = true }
 httpx = { version = ">=0.23.3", optional = true }
+dnacentersdk = "^2.5.6"
```
Comment (Kircheneer) [inline pyproject.toml:53]: I think this should be `optional = true`, also, is the actual minimum version 2.5.6 or can we got a bit lower?

File: nautobot_ssot/integrations/dna_center/diffsync/models/base.py (hunk 3 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/dna_center/diffsync/models/base.py:18]: I think I had at one point had Building as a child but removed that relationship. I should probably just remove the buildings attribute.

File: nautobot_ssot/integrations/dna_center/constants.py (hunk 1 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/dna_center/constants.py:8]: Was waiting to have a new netutils version released as we'll need to pin the version in the dependencies.

File: docs/admin/integrations/dna_center_setup.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+Before configuring the integration, please ensure, that `nautobot-ssot` app was [installed with Cisco DNA Center integration extra dependencies](../install.md#install-guide).
+
+```shell
+pip install nautobot-ssot[dna_center]
+```
+
+## Configuration
+
+Connecting to a DNA Center instance is handled through the Nautobot Controller object. There is an expectation that you will create an ExternalIntegration with the requisite connection information for your DNA Center instance attached to that Controller object. When running the Sync Job you will specify which DNA Center Controller instance you wish to synchronize with.  Other behaviors for the integration can be controlled with the following settings:
```
[the remark this answers is not in the document]
Comment (jdrew82) [reply docs/admin/integrations/dna_center_setup.md:15]: Done.

File: nautobot_ssot/integrations/dna_center/constants.py (hunk 1 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/dna_center/constants.py:8]: New release has been cut so I've updated the integration to use the mapping from there.

File: pyproject.toml (hunk 5 above)
Comment (jdrew82) [reply pyproject.toml:53]: Corrected the dependency to be optional.

File: nautobot_ssot/integrations/dna_center/diffsync/adapters/nautobot.py (hunk 2 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/dna_center/diffsync/adapters/nautobot.py:90]: I could see a select_related() making the query a lil more efficient so I've added that.

File: nautobot_ssot/integrations/dna_center/utils/dna_center.py (hunk 4 above)
Comment (jdrew82) [reply nautobot_ssot/integrations/dna_center/utils/dna_center.py:77]: Changes have been made.

File: pyproject.toml [hunk 7]
```diff
@@ -52,12 +52,14 @@ six = { version = ">=1.13.0", optional = true }
 httpx = { version = ">=0.23.3", optional = true }
 # Used by the Itential SSoT as a retry mechanism for HTTP failures in the AutomationGatewayClient.
 retry = "^0.9.2"
+# used for DNA Center integration
+dnacentersdk = "^2.5.6"
```
Comment (Kircheneer) [inline pyproject.toml:56]: I see its grouped now, but shouldn't it be

```suggestion
dnacentersdk = { version = "^2.5.6", optional = true }
```

Comment (jdrew82) [reply pyproject.toml]: You are correct. I've fixed that.

**Review by Kircheneer — APPROVED** [review]
LGTM

**Pull request merged**